api security measures

msadasivam opened the pull request:
What's changed, or what was fixed?

Add API security measures 
- Validate token using a ServiceCallout to apigee.googleapis.com. This ensures the user has access to the KVM.
- Validate org and env for each call so the call is ensured to come to the right org, env that was intended.
- Rename kvms to keyvaluemaps in the resource path to align with the mgmt API
- Spike Arrest

**Fixes:** #issue

- [x] I have run all the tests locally and they all pass.
- [x] I have followed the relevant style guide for my changes.

**CC:** @apigee-devrel-reviewers

danistrebel commented:
/gcbrun

danistrebel commented on references/kvm-admin-api/apiproxy/policies/AM-RemoveAuthzHeader.xml:
```diff
@@ -0,0 +1,23 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<AssignMessage continueOnError="false" enabled="true" name="AM-RemoveAuthzHeader">
```
We try to keep the boilerplate to a minimum and remove all default attributes and XML entries whenever possible. Can you remove `continueOnError="false" enabled="true"`?

danistrebel commented on references/kvm-admin-api/apiproxy/policies/RF-401.xml:
```diff
@@ -0,0 +1,26 @@
+ limitations under the License.
+-->
+
+<RaiseFault continueOnError="false" enabled="true" name="RF-401">
```
trivial: remove `continueOnError="false" enabled="true"`

danistrebel commented on references/kvm-admin-api/apiproxy/policies/SpikeArrest.xml:
```diff
@@ -0,0 +1,22 @@
+ limitations under the License.
+-->
+
+<SpikeArrest continueOnError="false" enabled="true" name="SpikeArrest">
```
Trivial for consistency:

rename: SA-SpikeArrest

remove: 
* `continueOnError="false" enabled="true"`
*  DisplayName
*  Properties

danistrebel commented on references/kvm-admin-api/apiproxy/policies/ValidateToken.xml:
```diff
@@ -0,0 +1,35 @@
+ limitations under the License.
+-->
+
+<ServiceCallout continueOnError="false" enabled="true" name="ValidateToken">
```
Trivial:

rename: SC-ValidateToken

remove:
* `continueOnError="false" enabled="true"`
*  DisplayName
*  Properties

danistrebel reviewed (changes requested):
Thanks Madhan, this is fantastic!!

For consistency I'd ask you to take a look at the trivial review comments.
Also because this breaks compatibility with Edge, I'd ask you to comment out the `pipeline.sh` steps such that we can get this in and figure out how to run this in the testbed for Apigee X separately.

msadasivam commented:
Please review, changes made are below:
- Used testIamPermissions for perms check
- Updated key attribute from name to key
- Added negative tests and updated basepath in test harness
- Updated doc

danistrebel commented on references/kvm-admin-api/apiproxy/targets/default.xml:
```diff
@@ -0,0 +1,32 @@
+        <Request/>
+        <Response>
+            <Step>
+                <Name>RF-403</Name>
```
This is most likely not enough. The way the `testIAMPermissions` check is implemented it always returns a 200 even if the permission is not available. E.g. with a read only service account I can list the KVMs

```sh
curl -H "Authorization: Bearer $TOKEN" -X POST "https://apigee.googleapis.com/v1/organizations/$APIGEE_ORG/environments/$APIGEE_ENV\:testIamPermissions" -H "Content-Type: application/json" -d "{\"permissions\": [\"apigee.keyvaluemaps.list\"]}"    
{
  "permissions": [
    "apigee.keyvaluemaps.list"
  ]
}
```
but I cannot create them:

```sh
curl -H "Authorization: Bearer $TOKEN" -X POST "https://apigee.googleapis.com/v1/organizations/$APIGEE_ORG/environments/$APIGEE_ENV\:testIamPermissions" -H "Content-Type: application/json" -d "{\"permissions\": [\"apigee.keyvaluemaps.create\"]}"     
{}
```

Both calls return a 200 though.

msadasivam replied:
You are right. I did overlook this key aspect. Fixed.

danistrebel commented on references/kvm-admin-api/test/integration/features/support/init.js:
```diff
@@ -21,8 +21,12 @@ before(function () {
     "https",
     process.env.APIGEE_ORG + "-" + process.env.APIGEE_ENV +
-    ".apigee.net/kvm-admin/v1"
+    ".apigee.net/kvm-admin/v1/organizations/" +
```
The edge hostname will not work with the GCP token validation. We should use something like an `env.APIGEE_HOSTNAME` variable instead.

msadasivam replied:
Fixed

danistrebel commented on references/kvm-admin-api/apiproxy/resources/jsc/JS-IAMPermission.js:
```diff
@@ -0,0 +1,14 @@
+else if(reqMethod === "DELETE")
+    iamPermission = "delete";
+else
+    iamPermission = "noop";
```
note:`apigee.keyvaluemaps.noop` will always fail with:

```json
{
  "error": {
    "code": 400,
    "message": "Permission apigee.keyvaluemaps.noop is not valid for this resource.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.RequestInfo",
        "requestId": "xxxx"
      }
    ]
  }
}
```

msadasivam replied:
Any unrecognized verb will trigger the 404 fallback in conditional flows. 
The noop is a little bit of a paranoid measure for future extensions. The block is unreachable for PUT requests.

laughingbiscuit commented:
thanks for the fixes madhan - reviewing again now!

laughingbiscuit commented:
/gcbrun

laughingbiscuit commented:
lgtm - will just wait for @danistrebel for resolve the last requested change

danistrebel reviewed (approved):
Thanks a lot for this PR @msadasivam. This is fantastic. LGTM! (validated on Apigee X)

danistrebel commented:
/gcbrun
